refactor: edit EFS UX to bool/struct, not string/struct

bvtujo opened the pull request:
<!-- Provide summary of changes -->
This PR adds the ability to specify managed EFS as a bool value: 

```yaml
storage:
  volumes:
    mySampleVolume:
      efs: true
      path: /mount/efs
    myManagedVolumeWithConfig:
      efs:
        uid: 100000
        gid: 1000000
      path: /mount/configured-efs
    myEphemeralVolume:
      efs: false
      path: /mount/scratch
    myCustomEFS:
      efs:
        id: fs-1234
        auth:
          access_point_id: fsap-1234
          iam: true
  ```

<!-- Issue number, if available. E.g. "Fixes #31", "Addresses #42, 77" -->

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

bvtujo committed:
chore: add logic to only update env if requires EFS

bvtujo committed:
chore: add custom unmarshaler to EFS

bvtujo committed:
chore: enable managed EFS

bvtujo committed:
chore: validate/autogenerate UID/GID

bvtujo committed:
chore: fix staticcheck error

bvtujo committed:
chore: add tests and fix failing integ tests

bvtujo committed:
chore: update templates to use EnvController outputs

bvtujo committed:
chore: fix cloudformation errors

bvtujo committed:
chore: add metadata to EFS filesystem and access point

bvtujo committed:
test: add integ tests for managed EFS

bvtujo committed:
chore: add go.sum

bvtujo committed:
Merge branch 'mainline' into managed-efs

bvtujo committed:
chore: add managed efs to scheduled jobs

bvtujo committed:
refactor: use bool/struct unmarshaler instead of string/struct for EFS

bvtujo committed:
chore: remove empty volume for integ test consistency

bvtujo committed:
chore: remove wlType from required input to convertStorageOpts

bvtujo committed:
chore: remove unused dirName function

bvtujo committed:
chore: address feedback

Add comment to avoid unnecessary safeguard for convertEFSConfig
Add comment for public function.

bvtujo committed:
chore: remove deadcode

bvtujo committed:
chore: address nits

bvtujo committed:
chore: add comment to EFSVolumeConfiguration

bvtujo committed:
chore: remove error from FSID

bvtujo committed:
chore: change quote style in error

Co-authored-by: Penghao He <penghaoh@amazon.com>

bvtujo committed:
Merge branch 'mainline' into managed-efs

bvtujo committed:
Merge branch 'managed-efs' into efs-refactor

bvtujo committed:
chore: address feedback on wrapped error messages

bvtujo committed:
Merge branch 'mainline' into efs-refactor

bvtujo committed:
Merge branch 'mainline' into efs-refactor

bvtujo committed:
Merge branch 'mainline' into efs-refactor

efekarakus commented on internal/pkg/deploy/cloudformation/stack/transformers.go:
```diff
@@ -35,16 +35,18 @@ const (
-	errNoFSID                      = errors.New(`volume field efs/id cannot be empty`)
-	errAcessPointWithRootDirectory = errors.New(`root directory must be empty or "/" when access point ID is specified`)
-	errAccessPointWithoutIAM       = errors.New(`"iam" must be true when access point ID is specified`)
+	errNoFSID                       = errors.New(`volume field efs/id cannot be empty`)
```
I wonder if we can bring more consistency into our error messages 💭 , what do you think of creating error categories and following a certain format per category?

For example,
```go
// Empty field errors.
var (
  errNoFSID = errors.New(`volume field "efs.id" cannot be empty`)
  errNoContainerPath = errors.New(`volume field "path" cannot be empty`)
)

// Conditional errors.
var (
   errAccessPointWithRootDirectory = errors.New(`volume field "efs.root_dir" must be empty or "/" when "access_point_id" is specified`)
)
```

efekarakus commented on internal/pkg/deploy/cloudformation/stack/transformers.go:
```diff
@@ -265,25 +253,23 @@ func convertEFSPermissions(input map[string]manifest.Volume) ([]*template.EFSPer
 		if volume.EFS.UseManagedFS() {
 			continue
 		}
+		// We're not using a managed volume and the EFS config is empty.
```
Is this a valid scenario? should we be throwing an error instead?

bvtujo replied:
Empty volumes are totally valid! We just don't need to add EFS permissions for them.

bvtujo replied:
This is now caught by `validateEFSConfiguration`.

efekarakus commented on internal/pkg/deploy/cloudformation/stack/transformers.go:
```diff
@@ -247,10 +238,7 @@ func convertMountPoints(input map[string]manifest.Volume) ([]*template.MountPoin
-		if err != nil {
-			return nil, err
-		}
+		mp := convertMountPoint(aws.String(name), volume.ContainerPath, volume.ReadOnly)
```
nit: this is a preference, but I really like the [inline variable refactor](https://refactoring.com/catalog/inlineVariable.html) to reduce the size of the fn body
```suggestion
		output = append(output,  convertMountPoint(aws.String(name), volume.ContainerPath, volume.ReadOnly))
```

efekarakus commented on internal/pkg/deploy/cloudformation/stack/transformers.go:
```diff
@@ -355,14 +326,24 @@ func convertVolumes(input map[string]manifest.Volume) ([]*template.Volume, error
 		//      data between containers.
-		if volume.EFS != nil && volume.EFS.UseManagedFS() {
+
+		// If EFS is not configured, just add the name to create an empty volume and continue.
+		if volume.EFS == nil || volume.EFS.EmptyVolume() {
```
Does this mean if someone specifies:
```yaml
volumes:
  myFS:
    efs:
```
We will create a scratch volume instead of an efs fs?

bvtujo replied:
ooh, yes. good catch.

efekarakus commented on internal/pkg/deploy/cloudformation/stack/workload.go:
```diff
@@ -264,6 +264,134 @@ func validatePath(input string, maxLength int) error {
+
+func validateVolume(name string, in manifest.Volume) error {
+	if err := validateMountPointConfig(in); err != nil {
+		return fmt.Errorf("validate container configuration for volume %s: %w", name, err)
```
nit: can we remove wrappers if we are calling private functions since we already return clear error messages from them

efekarakus commented on internal/pkg/manifest/storage.go:
```diff
@@ -49,15 +51,15 @@ func (e *EFSVolumeConfiguration) IsEmpty() bool {
-type EFSConfigOrID struct {
-	Config EFSVolumeConfiguration
-	ID     string
+// EFSConfigOrBool contains custom unmarshaling logic for the `efs` field in the manifest.
+type EFSConfigOrBool struct {
+	Config  EFSVolumeConfiguration
+	Enabled *bool
```
Recently with the spot feature we introduced this pattern to handle "advanced" configs:
https://github.com/aws/copilot-cli/blob/2c8bbc122130cb3b6035768d3edc6613a57bb861/internal/pkg/manifest/svc.go#L117-L122

What do you think of refactoring like this:
```suggestion
// EFS holds the configuration for the `efs` field.
type EFS struct {
	Advanced  EFSVolumeConfiguration
	Enabled  *bool
```

efekarakus commented on internal/pkg/manifest/storage.go:
```diff
@@ -69,50 +71,54 @@ func (e *EFSConfigOrID) UnmarshalYAML(unmarshal func(interface{}) error) error {
-	// If config is empty, use ID; otherwise ignore string and use struct.
-	if e.Config.IsEmpty() {
-		return aws.String(fromID)
+func (e *EFSVolumeConfiguration) EmptyBYOConfig() bool {
```
can we add doc comments to these exported methods :)

efekarakus commented on internal/pkg/manifest/storage.go:
```diff
@@ -69,50 +71,54 @@ func (e *EFSConfigOrID) UnmarshalYAML(unmarshal func(interface{}) error) error {
+}
+
+func (e *EFSConfigOrBool) EmptyVolume() bool {
+	// Respect Bool value first: return true if Enabled is false; false if true.
+	if e.Enabled != nil {
+		return !aws.BoolValue(e.Enabled)
```
Why are we flipping the boolean here? I think I'm not understanding what we mean by empty here

Lou1415926 replied:
nit:  ⬇️ might be easier to see the intention?
```suggestion
		return aws.BoolValue(e.Enabled) == false
```

efekarakus reviewed (commented):
Nice I really like these new validate functions!

Lou1415926 commented on internal/pkg/deploy/cloudformation/stack/transformers.go:
```diff
@@ -355,14 +326,24 @@ func convertVolumes(input map[string]manifest.Volume) ([]*template.Volume, error
+			continue
 		}
+
+		// If
```
```suggestion
```

Lou1415926 commented on internal/pkg/manifest/storage.go:
```diff
@@ -69,50 +71,54 @@ func (e *EFSConfigOrID) UnmarshalYAML(unmarshal func(interface{}) error) error {
+}
+
+func (e *EFSConfigOrBool) EmptyVolume() bool {
+	// Respect Bool value first: return true if Enabled is false; false if true.
```
I feel like we can have a more descriptive comment, for example:
```suggestion
	// Respect Bool value first: with Enabled being false, we don't want a managed EFS, hence creating an empty volume.
```
Or just 
```suggestion
	// Respect Bool value first.
```

Lou1415926 commented on internal/pkg/deploy/cloudformation/stack/transformers.go:
```diff
@@ -265,25 +253,23 @@ func convertEFSPermissions(input map[string]manifest.Volume) ([]*template.EFSPer
 		if volume.EFS.UseManagedFS() {
 			continue
 		}
+		// We're not using a managed volume and the EFS config is empty.
```
Should we move this comment to 3 lines above and say something more descriptive, such as  "We only need to add additional permissions in the case of BYO EFS system".

Lou1415926 reviewed (commented):
Looks great! Just a few nits on comment since this is a complicated feature and we might need the comments to understand some of the logic.

bvtujo committed:
Update internal/pkg/deploy/cloudformation/stack/transformers.go

Co-authored-by: Wanxian Yang <79273084+Lou1415926@users.noreply.github.com>

bvtujo committed:
chore: rename Config to Advanced in EFS

bvtujo committed:
chore: organize stack package

Move validators to a new file, validate.go.
Cluster error messages by category.

bvtujo committed:
chore: refactor tests; make sure empty 'efs' throws an error

bvtujo committed:
chore: move EmptyVolume func to Volume struct

bvtujo committed:
chore: use methods instead of direct logical comparisons"

bvtujo committed:
Merge branch 'mainline' into efs-refactor

bvtujo commented on internal/pkg/manifest/storage.go:
```diff
@@ -67,52 +82,52 @@ func (e *EFSConfigOrID) UnmarshalYAML(unmarshal func(interface{}) error) error {
+// This function is useful for checking that the EFS config has been intentionally turned off
+// and whether we should ignore any values of the struct which have been populated erroneously.
+func (e *EFSConfigOrBool) Disabled() bool {
+	if e.Enabled != nil && aws.BoolValue(e.Enabled) == false {
```
```suggestion
	if e.Enabled != nil && !aws.BoolValue(e.Enabled) {
```

bvtujo commented on internal/pkg/deploy/cloudformation/stack/validate.go:
```diff
@@ -0,0 +1,191 @@
+	}
+
+	// UID and GID are mutually exclusive with any other fields.
+	if in.EFS.Advanced.EmptyBYOConfig() == false && !in.EFS.Advanced.EmptyUIDConfig() {
```
```suggestion
      // lint:ignore S1002 we want to clarify the intent here by comparing to a boolean.
	if in.EFS.Advanced.EmptyBYOConfig() == false && !in.EFS.Advanced.EmptyUIDConfig() {
```

bvtujo committed:
chore: fix staticcheck errors

bvtujo committed:
chore: de Morgan's law refactor

mergify[bot] committed:
Merge branch 'mainline' into efs-refactor
